tomcat-gcp-lib/src/main/java/com/google/cloud/runtimes/tomcat/session/DatastoreStore.java

```diff
@@ -150,24 +159,44 @@ public int getSize() throws IOException {
   @Override
   public Session load(String id) throws ClassNotFoundException, IOException {
     log.debug("Session " + id + " requested");
-    StandardSession session = null;
+    TraceContext context = startSpan("Loading session");
+    Key sessionKey = newKey(id);
 
-    Entity sessionEntity = datastore.get(keyFactory.newKey(id));
-
-    if (sessionEntity != null) {
-      try (InputStream fis = sessionEntity.getBlob("content").asInputStream();
-          ObjectInputStream ois = getObjectInputStream(fis)) {
-        session = (StandardSession) manager.createEmptySession();
-        session.readObjectData(ois);
-        session.setManager(manager);
-      }
-    }
+    DatastoreSession session = deserializeSession(sessionKey);
 
+    endSpan(context);
     log.debug("Session " + id + " loaded");
     return session;
   }
 
   /**
+   * Create a new session usable by Tomcat, from a serialized session in a Datastore Entity.
+   * @param sessionKey The key associated with the session metadata and attributes.
+   * @return A new session containing the metadata and attributes stored in the entity.
+   * @throws ClassNotFoundException Thrown if a class serialized in the entity is not available in
+   *                                this context.
+   * @throws IOException Thrown when an error occur during the deserialization.
+   */
+  private DatastoreSession deserializeSession(Key sessionKey)
+      throws ClassNotFoundException, IOException {
+    TraceContext loadingSessionContext = startSpan("Fetching the session from Datastore");
+    Iterator<Entity> entities = datastore.run(Query.newEntityQueryBuilder()
+        .setKind(sessionKind)
+        .setFilter(PropertyFilter.hasAncestor(sessionKey))
+        .build());
+    endSpan(loadingSessionContext);
+
+    DatastoreSession session = null;
+    if (entities.hasNext()) {
```

Comment:
What happens if more than one entity is returned? Should we throw an exception? Log a warning message?

Reply:
It is actually expected to return multiple entities (one for the session metadata and one for each attribute) if the metadata is not present an Exception will be thrown during the deserialization.

This test check if a session with this Id already exists (at least one entity is fetched), if not then null is returned and the Manager will create a new session locally.